src/Features/LanguageServer/Protocol/LspMiscellaneousFilesWorkspace.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Features.Workspaces;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.LanguageServer
+{
+    /// <summary>
+    /// Defines a default workspace for opened LSP files that are not found in any
+    /// workspace registered by the <see cref="ILspWorkspaceRegistrationService"/>.
+    /// If a document added here is subsequently found in a registered workspace, 
+    /// the document is removed from this workspace.
+    /// 
+    /// Future work for this workspace includes supporting basic metadata references (mscorlib, System dlls, etc),
+    /// but that is dependent on having a x-plat mechanism for retrieving those references from the framework / sdk.
+    /// </summary>
+    internal class LspMiscellaneousFilesWorkspace : Workspace
+    {
+        private const string LspMiscellaneousFilesWorkspaceKind = nameof(LspMiscellaneousFilesWorkspaceKind);
+
+        private static readonly LanguageInformation s_csharpLanguageInformation = new(LanguageNames.CSharp, ".csx");
+        private static readonly LanguageInformation s_vbLanguageInformation = new(LanguageNames.VisualBasic, ".vbx");
+
+        private static readonly Dictionary<string, LanguageInformation> s_extensionToLanguageInformation = new()
+        {
+            { ".cs", s_csharpLanguageInformation },
+            { ".csx", s_csharpLanguageInformation },
+            { ".vb", s_vbLanguageInformation },
+            { ".vbx", s_vbLanguageInformation },
+        };
+
+        private readonly ILspLogger _logger;
+
+        public LspMiscellaneousFilesWorkspace(ILspLogger logger) : base(MefHostServices.DefaultHost, LspMiscellaneousFilesWorkspaceKind)
```

Comment:
Do we not have an existing MEF host we can reuse?

Reply:
we have various workspaces registered, which one do we use though?  The misc files workspace in VS uses VS workspace services, but that doesn't exist outside of VS.  We could pick one of the LSP registered workspaces, but which one do we pick?  I think just doing this makes the most sense.